runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/fn/control/ProcessRemoteBundleOperation.java

```diff
@@ -100,6 +218,80 @@ public void finish() throws Exception {
       } catch (Exception e) {
         throw new RuntimeException("Failed to finish remote bundle", e);
       }
+
+      try {
+        // In Batch, the input watermark advances from -inf to +inf once all elements are processed.
+        // For timers, this means we only fire them at the end of the bundle.
+        fireTimers();
+      } catch (Exception e) {
+        throw new RuntimeException("Failed to finish remote bundle", e);
+      }
+    }
+  }
+
+  // Recursively fires all scheduled timers from the SDK Harness.
+  private void fireTimers() throws Exception {
+    DataflowExecutionContext.DataflowStepContext stepContext =
+        executionContext.getStepContext((DataflowOperationContext) this.context).namespacedToUser();
+
+    TimerInternals.TimerData timerData = stepContext.getNextFiredTimer(GlobalWindow.Coder.INSTANCE);
+    while (timerData != null) {
+      StateNamespaces.WindowNamespace windowNamespace =
+          (StateNamespaces.WindowNamespace) timerData.getNamespace();
+      BoundedWindow window = windowNamespace.getWindow();
+
+      WindowedValue<KV<Object, Timer>> timerValue =
+          WindowedValue.of(
+              KV.of(
+                  timerIdToKey.get(timerData.getTimerId()),
+                  Timer.of(timerData.getTimestamp(), timerIdToPayload.get(timerData.getTimerId()))),
```

Comment:
This is nice. Per our in-person conversation the timers already have separate timestamps and delivery instructions. They are `WindowedValue.of(KV.of(id, delivery), timestamp, ...)`

Or anyhow that's how I would interpret this construction here. Is that right?